src/overseer/core.clj

```diff
@@ -1,97 +1,62 @@
-   txn data merged onto each txn"
-  [job-types tx]
-  (zipmap
-    (map identity job-types)
-    (map #(job-txn % tx) job-types)))
-
-(defn job-dep-edges
-  "Construct a list of txns to of the graph edges, i.e marking
-   job dependencies"
-  [graph jobs-by-type]
-  (for [[job deps] graph
-        dep deps]
-    {:db/id (get-in jobs-by-type [job :db/id])
-     :job/dep (get-in jobs-by-type [dep :db/id])}))
-
-(defn ->job-entity [db job-id]
-  {:pre [job-id]}
-  (d/pull db '[:*] [:job/id job-id]))
-
-(defn ent-dependents
-  "Find all jobs entities that depend on ent"
-  [db ent]
-  (->> (d/q '[:find ?dep
-              :in $ ?ent
-              :where [?dep :job/dep ?ent]]
-            db
-            ent)
-       (map first)
-       (set)))
-
-(defn transitive-dependents [db job-id]
-  "Returns a set of job IDs that transitively depend upon given job ID
-   Basically recursive breadth-first graph traversal of the job graph."
-  (let [job-ent-id (:db/id (->job-entity db job-id))]
-    (assert job-ent-id)
-    (loop [all-dependents #{}
-           visited #{}
-           to-visit #{job-ent-id}]
-      (if (empty? to-visit)
-        (set (map #(:job/id (d/entity db %)) all-dependents))
-        (let [ent (first to-visit)
-              dependents (ent-dependents db ent)
-              all-dependents' (set/union all-dependents dependents)
-              visited' (conj visited ent)
-              to-visit' (set/difference
-                          (set/union to-visit dependents)
-                          visited')]
-          (recur all-dependents' visited' to-visit'))))))
-
-(defn status-txn
-  "Construct a single job update status txn"
-  [{:keys [overseer/status overseer/failure]} job-id]
-  (let [base-txn {:db/id [:job/id job-id]
-                  :job/status status}]
-    (if failure
-      (assoc base-txn :job/failure (pr-str failure))
-      base-txn)))
-
-(defn update-job-status-txns
-  "Construct a seq of txns for updating a job's status
-   If job was aborted, then also abort all its dependents"
-  [db job-id {:keys [overseer/status] :as status-map}]
-  (let [job-ids (if (= :aborted status)
-                  (cons job-id (transitive-dependents db job-id))
-                  [job-id])]
-    (map (partial status-txn status-map) job-ids)))
+(defprotocol Store
+  (job-info [this job-id])
+
+  (reserve-job [this job-id])
+  (finish-job [this job-id])
+  (fail-job [this job-id failure])
+  (heartbeat-job [this job-id])
+  (abort-job [this job-id])
+  (reset-job [this job-id])
+
+  (jobs-ready [this])
+  (jobs-dead [this thresh])
+
+  (transact-graph [this graph]))
+
+(defn squuid []
+  (let [uuid (java.util.UUID/randomUUID)
+        time (System/currentTimeMillis)
+        secs (quot time 1000)
+        lsb (.getLeastSignificantBits uuid)
+        msb (.getMostSignificantBits uuid)
+        timed-msb (bit-or (bit-shift-left secs 32)
+                          (bit-and 0x00000000ffffffff msb))]
+    (java.util.UUID. timed-msb lsb)))
```

Comment:
Is it worth commenting on the what/why of squuids / delegating to the section in [Datomic Identity docs](http://docs.datomic.com/identity.html)? Is this an invented algorithm or one sourced from somewhere (is there a standard formula?) ? Is it worth just shelling out to Datomic's impl here (already present as a dep; obviously this ns is not otherwise tied to Datomic semantics) ?